src/timer_wrap.h

```diff
@@ -0,0 +1,87 @@
+#ifndef SRC_TIMER_WRAP_H_
+#define SRC_TIMER_WRAP_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include "memory_tracker.h"
+#include "env.h"
+#include "uv.h"
+
+#include <functional>
+
+namespace node {
+
+// Utility class that makes working with libuv timers a bit easier.
+class TimerWrap final : public MemoryRetainer {
+ public:
+  using TimerCb = std::function<void(void*)>;
+
+  TimerWrap(Environment* env, TimerCb fn, void* user_data);
+  TimerWrap(const TimerWrap&) = delete;
+
+  inline Environment* env() const { return env_; }
+
+  // Completely stops the timer, making it no longer usable.
+  void Stop(bool close = true);
+
+  // Starts / Restarts the Timer
+  void Update(uint64_t interval, uint64_t repeat = 0);
+
+  void Ref();
+
+  void Unref();
+
+  SET_NO_MEMORY_INFO();
+  SET_MEMORY_INFO_NAME(TimerWrap)
+  SET_SELF_SIZE(TimerWrap)
+
+ private:
+  static void TimerClosedCb(uv_handle_t* handle);
+  static void OnTimeout(uv_timer_t* timer);
+  ~TimerWrap() = default;
+
+  Environment* env_;
+  TimerCb fn_;
+  uv_timer_t timer_;
+  void* user_data_ = nullptr;
+
+  friend std::unique_ptr<TimerWrap>::deleter_type;
+};
+
+class TimerWrapHandle : public MemoryRetainer  {
+ public:
+  TimerWrapHandle(
+      Environment* env,
+      TimerWrap::TimerCb fn,
+      void* user_data = nullptr);
+
+  TimerWrapHandle(const TimerWrapHandle&) = delete;
+
+  ~TimerWrapHandle() { Stop(); }
+
+  void Update(uint64_t interval, uint64_t repeat = 0);
+
+  void Ref();
+
+  void Unref();
+
+  void Stop(bool close = true);
+
+  inline void MemoryInfo(node::MemoryTracker* tracker) const override {
+    if (timer_ != nullptr)
+      tracker->TrackField("timer", *timer_);
+  }
```

Comment:
This can’t be used as an inline method because it’s virtual (and only accessed through base class pointers), and it forces all users of this header to also include `memory_tracker-inl.h`, so I would move it into the `.cc` file.